eth/db/journal.py

```diff
@@ -24,237 +26,258 @@ class DeletedEntry:
 
     @property
-    def root_changeset_id(self) -> uuid.UUID:
+    def root_changeset_id(self) -> JournalDBCheckpoint:
         """
         Returns the id of the root changeset
         """
-        return first(self.journal_data.keys())
+        return first(self._journal_data.keys())
 
     @property
     def is_flattened(self) -> bool:
         """
         :return: whether there are any explicitly committed checkpoints
         """
-        return len(self.journal_data) < 2
+        return len(self._checkpoint_stack) < 2
 
     @property
-    def latest_id(self) -> uuid.UUID:
+    def last_changeset_index(self) -> JournalDBCheckpoint:
         """
         Returns the id of the latest changeset
         """
         # last() was iterating through all values, so first(reversed()) gives a 12.5x speedup
-        return first(reversed(self.journal_data.keys()))
-
-    @property
-    def latest(self) -> Dict[bytes, Union[bytes, DeletedEntry]]:
-        """
-        Returns the dictionary of db keys and values for the latest changeset.
-        """
-        return self.journal_data[self.latest_id]
-
-    @latest.setter
-    def latest(self, value: Dict[bytes, Union[bytes, DeletedEntry]]) -> None:
-        """
-        Setter for updating the *latest* changeset.
-        """
-        self.journal_data[self.latest_id] = value
-
-    def is_empty(self) -> bool:
-        return len(self.journal_data) == 0
+        # Interestingly, an attempt to cache this value caused a slowdown.
+        return first(reversed(self._journal_data.keys()))
 
-    def has_changeset(self, changeset_id: uuid.UUID) -> bool:
-        return changeset_id in self.journal_data
+    def has_changeset(self, changeset_id: JournalDBCheckpoint) -> bool:
+        # another option would be to enforce monotonically-increasing changeset ids, so we can do:
+        # checkpoint_idx = bisect_left(self._checkpoint_stack, changeset_id)
+        # (then validate against length and value at index)
+        return changeset_id in self._checkpoint_stack
 
-    def record_changeset(self, custom_changeset_id: uuid.UUID = None) -> uuid.UUID:
+    def record_changeset(
+            self,
+            custom_changeset_id: JournalDBCheckpoint = None) -> JournalDBCheckpoint:
         """
-        Creates a new changeset. Changesets are referenced by a random uuid4
+        Creates a new changeset. Changesets are referenced by a random int
         to prevent collisions between multiple changesets.
         """
         if custom_changeset_id is not None:
-            if custom_changeset_id in self.journal_data:
+            if custom_changeset_id in self._journal_data:
                 raise ValidationError(
                     "Tried to record with an existing changeset id: %r" % custom_changeset_id
                 )
             else:
                 changeset_id = custom_changeset_id
         else:
-            changeset_id = uuid.uuid4()
+            changeset_id = get_next_checkpoint_id()
```

Comment:
We used to call the recorded changes `changeset` and the ids referencing them `changeset_id`. The new number generator calls them `checkpoint_id` which introduces a slight inconsistency of terminology. I think we should make the following renames:

`get_next_checkpoint_id` -> `get_next_changeset_id`
`JournalDBCheckpoint` -> `JournalChangesetID` (notice the dropped `DB` because the class is also called `Journal` not `JournalDB`

We could also go the other way around and use `checkpoint` everywhere but I think I have a slight preference for `changeset` and `changeset_id` because it maps better to the actual concepts in my head. A `changeset`  (and `changeset_id`) implies that we are holding a *set*  of *changes* that we can reference by some assigned *id*  which is exactly what is happening.

Reply:
Okay, onboard with getting the naming unified, but with the opposite preference. Here's why checkpoint makes more sense to me:
- `checkpoint` is already an ID. So we are comparing `checkpoint` to `changeset_id`
- there are other places (like tests) that the name is shortened to `changeset` (presumably because it's shorter/easier to type)
- Dropping from `changeset_id` to `changeset` dramatically changes the meaning. I expect a `changeset` to be something like a `DBDiff`, a set of all the changes (between two checkpoints).
- committing or discarding a changeset (ID) seems like we're only committing or discarding the changes that happened in that specific changeset
- committing a checkpoint is more intuitively committing everything that happened since the checkpoint.
- put another way: it's like the changeset for every single changeset ID is changing every time you add or delete to the journal (committing the first changeset has a different meaning after every touch of the journal)

Re: dropping `DB`, all of the classes that refer to the checkpoint (like `VMState`, `AccountDB`, etc) are using the `JournalDB` class and theoretically don't know/care that the `Journal` implementation exists. Of course I get that it's strange that `Journal` would reference `JournalDB`. I guess theoretically there should be two checkpoint classes, one that would only be used internally. That's kind of more babysitting of mypy than I feel like doing. I've come close to dropping `Journal` altogether and inlining it to `JournalDB` since we're not using `Journal` elsewhere and it's so performance sensitive. So I really didn't feel like wrangling mypy if that's where we're headed.

Reply:
I just skimmed over this really fast, so no in-depth reply here. Just wanted to signal that I don't  have a strong preference here so whatever resolves the inconsistency is :+1: for me.